### PR TITLE
fix: Limit content length size when downloading link preview (WEBAPP-4422, WEBAPP-5341)

### PR DESCRIPTION
**Pull request by atomrc**
Previous strategy was:

- do HEAD request and read the content of the response ;
- if the content type doesn't match throw error ;
- else send another request via `openGraph`.

New strategy:

- do full request ;
- when the headers arrive (before any content is downloaded), check that `content-type` is of type `text/html` ;
- start counting downloaded bytes ;
- if the number of bytes exceed the limit, we fail ;
- if not we call the `parse` method of `openGraph` only on the `<head>` of the document.

### CODE REVIEW

File: electron/js/lib/openGraph.js [hunk 1]
```diff
@@ -41,21 +42,37 @@ const fetchImageAsBase64 = url => {
...
+        throw new Error(`Unhandled format for open graph generation ('${contentType}' of size '${contentLength}')`);
+        getContentRequest.abort();
+      }
+    });
+
+    let requestCurrentSize = 0;
+    getContentRequest.on('data', buffer => {
+      requestCurrentSize += buffer.length;
+      if (requestCurrentSize > CONTENT_TYPE_LIMIT) {
+        throw new Error(`File size too big for open graph generation ('${contentLength}')`);
```
Comment (pwnsdx) [inline electron/js/lib/openGraph.js]: Error throwing must come after aborting the request, they act as a return.

File: electron/js/lib/openGraph.js [hunk 2]
```diff
@@ -41,21 +42,37 @@ const fetchImageAsBase64 = url => {
...
-  });
+    getContentRequest.on('response', ({headers}) => {
+      const contentType = headers['content-type'] || '';
+      const contentLength = parseInt(headers['content-length'] || '0', 10);
+
+      const isHtmlContentType = contentType.match(/.*text\/html/);
+      const isTooLarge = contentLength > CONTENT_TYPE_LIMIT;
+
+      if (!isHtmlContentType || isTooLarge) {
+        throw new Error(`Unhandled format for open graph generation ('${contentType}' of size '${contentLength}')`);
```
Comment (pwnsdx) [inline electron/js/lib/openGraph.js]: Error throwing must come after aborting the request, they act as a return.

File: electron/js/lib/openGraph.js (hunk 1 above)
Comment (gregor) [inline electron/js/lib/openGraph.js]: Let's try to not through but parse partial downloads. Open graph data can be found in `<meta>` tags within then `<head>`. Thus we can set the limit much, much lower and even check for the closing head tag.

File: electron/js/lib/openGraph.js [hunk 3]
```diff
@@ -41,21 +42,37 @@ const fetchImageAsBase64 = url => {
...
-      return new Promise((resolve, reject) => openGraph(url, (error, meta) => (error ? reject(error) : resolve(meta))));
-    }
-    throw new Error('Unhandled format for open graph generation');
-  });
+    getContentRequest.on('response', ({headers}) => {
+      const contentType = headers['content-type'] || '';
+      const contentLength = parseInt(headers['content-length'] || '0', 10);
+
+      const isHtmlContentType = contentType.match(/.*text\/html/);
+      const isTooLarge = contentLength > CONTENT_TYPE_LIMIT;
```
Comment (pwnsdx) [inline electron/js/lib/openGraph.js]: If you really want to check the content type and its length then you need to restore the HEAD request otherwise those checks only happen at the end.

File: electron/js/lib/openGraph.js [hunk 4]
```diff
@@ -41,21 +42,37 @@ const fetchImageAsBase64 = url => {
...
 const fetchOpenGraphData = url => {
-  const isHtmlContentTypePromise = new Promise(resolve => {
-    const headerRequest = request.head(url);
-    headerRequest.on('response', response => {
-      const isHtmlContentType =
-        response.headers['content-type'] && response.headers['content-type'].match(/.*text\/html/);
-      resolve(isHtmlContentType);
+  const CONTENT_TYPE_LIMIT = 1e7; // 10MB
+  const parsedUrl = urlUtil.parse(url);
+  const normalizedUrl = parsedUrl.protocol ? parsedUrl : urlUtil.parse(`http://${url}`);
```
Comment (pwnsdx) [inline electron/js/lib/openGraph.js]: Request should not happen if no protocol are set.

**Comment by pwnsdx** [comment]
Downloading images should also be subject to size limitation: https://github.com/wireapp/wire-desktop/pull/1671/files#diff-5d44f5d58de8cfcdc007eebe371758a9R33

Unifying the two functions in one would be ideal.

File: electron/js/lib/openGraph.js (hunk 1 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: The strategy changed with my last commit

File: electron/js/lib/openGraph.js (hunk 4 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: That's actually a fix for a bug we have. That's expected behavior

File: electron/js/lib/openGraph.js (hunk 3 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: The `on('response')` event is actually called before any content is downloaded (basically when the headers are here)

File: electron/js/lib/openGraph.js [hunk 5]
```diff
@@ -41,21 +42,38 @@ const fetchImageAsBase64 = url => {
...
+        throw new Error(`Unhandled format for open graph generation ('${contentType}')`);
+      }
+    });
+
+    let requestCurrentSize = 0;
+    let partialBody = '';
+    getContentRequest.on('data', buffer => {
+      const chunk = buffer.toString();
+      requestCurrentSize += buffer.length;
+      partialBody += chunk;
```
[the remark this answers is not in the document]
Comment (atomrc) [reply electron/js/lib/openGraph.js]: Humm I would argue that this is not the same thing. `buffer.length` is the number of bytes while `body.length` is the number of characters (that could differ because of unicode).

File: electron/js/lib/openGraph.js (hunk 3 above)
Comment (pwnsdx) [reply electron/js/lib/openGraph.js]: Headers are part of the downloaded content. I think you need a HEAD request.

Comment (pwnsdx) [reply electron/js/lib/openGraph.js]: Another approach if we don't want to do 2 requests would be extension blacklisting but it's far from perfect.

File: electron/js/lib/openGraph.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (pwnsdx) [reply electron/js/lib/openGraph.js]: Exact but it does not need to be that precise ;)

[the remark this answers is not in the document]
Comment (pwnsdx) [reply electron/js/lib/openGraph.js]: (If you feel you need to keep this precision then I would just suggest to rename `requestCurrentSize` to `responseCurrentSize`)

[the remark this answers is not in the document]
Comment (atomrc) [reply electron/js/lib/openGraph.js]: The number of characters is good enough indeed

File: electron/js/lib/openGraph.js (hunk 3 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: I checked, this function is called before any `data` event occurs. We are pretty safe on that side :+1:

Comment (atomrc) [reply electron/js/lib/openGraph.js]: You can find the lyfecycle of a request here https://nodejs.org/api/http.html#http_http_request_options_callback

Actually if the body is empty, no `data` event will be emitted.

File: electron/js/lib/openGraph.js [hunk 6]
```diff
@@ -28,34 +29,63 @@ const bufferToBase64 = (buffer, mimeType) => {
 };
 
 const fetchImageAsBase64 = url => {
+  const IMAGE_SIZE_LIMIT = 5e6; // 5MB
```
Comment (lipis) [inline electron/js/lib/openGraph.js]: 👍

File: electron/js/lib/openGraph.js [hunk 7]
```diff
@@ -28,34 +29,63 @@ const bufferToBase64 = (buffer, mimeType) => {
...
 };
 
 const fetchOpenGraphData = url => {
-  const isHtmlContentTypePromise = new Promise(resolve => {
-    const headerRequest = request.head(url);
-    headerRequest.on('response', response => {
-      const isHtmlContentType =
-        response.headers['content-type'] && response.headers['content-type'].match(/.*text\/html/);
-      resolve(isHtmlContentType);
+  const CONTENT_SIZE_LIMIT = 1000000; // ~1MB
```
Comment (lipis) [inline electron/js/lib/openGraph.js]: `1e6`?

File: electron/js/lib/openGraph.js [hunk 8]
```diff
@@ -28,34 +29,63 @@ const bufferToBase64 = (buffer, mimeType) => {
 };
 
 const fetchImageAsBase64 = url => {
+  const IMAGE_SIZE_LIMIT = 5e6; // 5MB
   return new Promise(resolve => {
-    request({encoding: null, url: encodeURI(url)}, (error, response, body) => {
+    const imageRequest = request({encoding: null, url: encodeURI(url)}, (error, response, body) => {
       if (!error && response.statusCode === 200) {
```
Comment (lipis) [inline electron/js/lib/openGraph.js]: can we start avoiding magic numbers?

maybe adding all the status codes in one file and have it like `status.OK`?

Could be part of another PR though..

File: electron/js/lib/openGraph.js (hunk 7 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: absolutely :)

File: electron/js/lib/openGraph.js (hunk 8 above)
Comment (atomrc) [reply electron/js/lib/openGraph.js]: Good point but, yeah, probably not the scope of this PR
